File: crates/inlaysql-ffi/examples/poc.php
<?php

/**
 * InlaySQL from PHP, through the C ABI — a proof of concept.
 *
 * This is the "SQLite-like adapter" direction: no server, the file opened
 * in-process through PHP's built-in FFI ext. It is what a `pdo_inlaysql`
 * extension would wrap, and what the Laravel driver in the docs would call.
 *
 * Build the library first (repo root):
 *   cargo build -p inlaysql-ffi --release
 *
 * Run this:
 *   php examples/ffi/poc.php [path/to/libinlaysql.dylib|.so|.dll] [db path]
 *
 * Requires PHP 7.4+ with the FFI extension (bundled, not enabled by default).
 * Preload/FPM needs ffi.enabled=preload with the definitions in a preloaded
 * file; the CLI here uses the simple embed mode.
 */

if ($argc < 2) {
    fwrite(STDERR, "usage: php poc.php <libinlaysql> [db-path]\n");
    fwrite(STDERR, "  <libinlaysql>  path to libinlaysql.dylib / .so / .dll\n");
    exit(1);
}

$lib = $argv[1];
$dbPath = $argv[2] ?? sys_get_temp_dir() . '/inlaysql-php-poc.inlay';

if (!file_exists($lib)) {
    fwrite(STDERR, "library not found: $lib\n");
    exit(1);
}

// The definitions mirror crates/inlaysql-ffi/include/inlaysql.h. Kept in one
// place here; an extension or packaged driver would ship them as a .h beside
// the library.
$ffi = FFI::cdef(<<<'C'
    typedef struct InlaysqlHandle InlaysqlHandle;

    InlaysqlHandle *inlaysql_open(const char *path);
    InlaysqlHandle *inlaysql_open_read_only(const char *path);
    void inlaysql_close(InlaysqlHandle *handle);

    int inlaysql_exec(InlaysqlHandle *handle,
                      const char *sql,
                      const char *params,
                      char **out_json);

    const char *inlaysql_last_error(void);
    void inlaysql_free_string(char *s);
    const char *inlaysql_version(void);
C, $lib);

/** Run one statement; returns the decoded JSON, or throws with the engine's error. */
function inlaysql_exec(object $ffi, object $handle, string $sql, array $params = []): array
{
    $paramsJson = $params === [] ? null : json_encode($params, JSON_UNESCAPED_SLASHES);
    $out = $ffi->new('char *');

    $code = $ffi->inlaysql_exec(
        $handle,
        $sql,
        $paramsJson,
        FFI::addr($out),
    );

    if ($code !== 0) {
        // `inlaysql_last_error` returns `const char*`; PHP's FFI surfaces a
        // plain `char*` return as a native string, so no FFI::string dance.
        $err = $ffi->inlaysql_last_error();
        $message = $code === 2
            ? 'bad handle'
            : (is_string($err) ? $err : FFI::string($err));
        throw new RuntimeException("InlaySQL error ($code): $message — while running: $sql");
    }

    // `out` however is a `char**` we own, so its pointee arrives as FFI\CData.
    $json = is_string($out) ? $out : FFI::string($out);
    $ffi->inlaysql_free_string($out);

    return json_decode($json, true, flags: JSON_THROW_ON_ERROR);
}

// `inlaysql_version` returns a static `const char*`; PHP's FFI surfaces a
// return of that type as a raw string, not a pointer — `FFI::string` is only
// for pointers we own an FFI\CData view of.
$version = $ffi->inlaysql_version();
echo "InlaySQL engine version $version\n";
echo "database: $dbPath\n\n";

// Open (creating if needed). Close is PHP-GC driven: when $handle dies the
// FFI handle is not automatically freed, so a real driver pairs this with an
// object that calls inlaysql_close in __destruct.
$handle = $ffi->inlaysql_open($dbPath);
if (FFI::isNull($handle)) {
    fwrite(STDERR, 'open failed: ' . FFI::string($ffi->inlaysql_last_error()) . "\n");
    exit(1);
}

try {
    $r = inlaysql_exec($ffi, $handle, 'CREATE TABLE IF NOT EXISTS docs (
        id INTEGER PRIMARY KEY,
        title TEXT,
        body TEXT
    )');
    echo "create:      {$r['kind']}\n";

    $r = inlaysql_exec($ffi, $handle, 'INSERT INTO docs (title, body) VALUES (?, ?)', [
        'Hello', 'from PHP over the C ABI — no server, no MySQL, the file is ours.',
    ]);
    echo "insert:      {$r['kind']}, {$r['rows']} row\n";

    $r = inlaysql_exec($ffi, $handle, 'SELECT id, title, body FROM docs ORDER BY id');
    echo "select:      {$r['rows'][0][0]} {$r['rows'][0][1]}\n";
    echo "             {$r['rows'][0][2]}\n";

    // Bound parameters carry the same rules as every other surface: strings,
    // numbers, null — and nested number arrays as vectors.
    $r = inlaysql_exec($ffi, $handle, 'SELECT COUNT(*) AS n FROM docs');
    echo "count:       {$r['rows'][0][0]}\n";

    // And errors surface the engine's own words:
    try {
        inlaysql_exec($ffi, $handle, 'SELECT * FROM no_such_table');
    } catch (RuntimeException $e) {
        echo "error path:  {$e->getMessage()}\n";
    }
} finally {
    $ffi->inlaysql_close($handle);
}

echo "\nOK — PHP drove the engine in-process through the C ABI.\n";

File: crates/inlaysql-ffi/examples/poc.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

# InlaySQL from Ruby, through the C ABI — a proof of concept.
#
# This is the "SQLite-like adapter" direction: no server, the file opened
# in-process through Ruby's `ffi` gem. It is the whole binding — nothing
# else is needed, which is the point.
#
# Build the library first (repo root):
#   cargo build -p inlaysql-ffi --release
#
# Run this:
#   gem install ffi          # the only gem; stdlib otherwise
#   ruby examples/poc.rb [path/to/libinlaysql.dylib|.so] [db path]

require "ffi"
require "json"
require "tmpdir"

# The binding. Fifteen lines — this is the whole thing, and the reason a C
# ABI is worth maintaining: every FFI language writes one of these and is
# done.
module InlaySQL
  LIB = ARGV[0] || File.expand_path(
    "../../../target/release/libinlaysql_ffi.#{RbConfig::CONFIG['SOEXT']}",
    __dir__
  )
  INLAYSQL_OK = 0
  INLAYSQL_ERR_BAD_HANDLE = 2

  module Native
    extend FFI::Library
    ffi_lib LIB

    attach_function :inlaysql_open, [:string], :pointer
    attach_function :inlaysql_open_read_only, [:string], :pointer
    attach_function :inlaysql_close, [:pointer], :void
    attach_function :inlaysql_exec, [:pointer, :string, :string, :pointer], :int
    attach_function :inlaysql_last_error, [], :string
    attach_function :inlaysql_free_string, [:pointer], :void
    attach_function :inlaysql_version, [], :string
  end

  def self.version
    Native.inlaysql_version
  end

  # A handle with a block form and a finalizer, so close is not a memory
  # leak waiting to happen. `exec` takes plain Ruby values: strings, ints,
  # floats, nil, arrays of numbers (vectors).
  class Database
    def initialize(path)
      @handle = Native.inlaysql_open(path)
      raise "open failed: #{Native.inlaysql_last_error}" if @handle.null?
    end

    def exec(sql, params = nil)
      out = FFI::MemoryPointer.new(:pointer)
      code = Native.inlaysql_exec(
        @handle, sql,
        params ? JSON.generate(params) : nil,
        out
      )
      case code
      when INLAYSQL_ERR_BAD_HANDLE then raise "bad handle"
      when INLAYSQL_OK
        result = JSON.parse(out.read_pointer.read_string)
        Native.inlaysql_free_string(out.read_pointer)
        result
      else
        raise "#{Native.inlaysql_last_error} — while running: #{sql}"
      end
    end

    def close
      Native.inlaysql_close(@handle)
      @handle = nil
    end
  end
end

lib_desc = InlaySQL::LIB
abort("library not found: #{lib_desc}\nbuild it: cargo build -p inlaysql-ffi --release") unless File.exist?(lib_desc)

db_path = ARGV[1] || File.join(Dir.tmpdir, "inlaysql-ruby-poc.inlay")

puts "InlaySQL engine version #{InlaySQL.version}"
puts "database: #{db_path}"
puts

db = InlaySQL::Database.new(db_path)
begin
  r = db.exec(<<~SQL)
    CREATE TABLE IF NOT EXISTS docs (
      id INTEGER PRIMARY KEY,
      title TEXT,
      body TEXT
    )
  SQL
  puts "create:      #{r['kind']}"

  r = db.exec("INSERT INTO docs (title, body) VALUES (?, ?)",
              ["Hello", "from Ruby over the C ABI — no server, the file is ours."])
  puts "insert:      #{r['kind']}, #{r['rows']} row"

  r = db.exec("SELECT id, title, body FROM docs ORDER BY id")
  puts "select:      #{r['rows'][0][0]} #{r['rows'][0][1]}"
  puts "             #{r['rows'][0][2]}"

  r = db.exec("SELECT COUNT(*) FROM docs")
  puts "count:       #{r['rows'][0][0]}"

  begin
    db.exec("SELECT * FROM no_such_table")
  rescue StandardError => e
    puts "error path:  InlaySQL error: #{e.message}"
  end
ensure
  db.close
end

puts
puts "OK — Ruby drove the engine in-process through the C ABI."
